tag_reader.go

```diff
@@ -71,12 +74,37 @@ func (r *tagReader) ReadAll() ([]*Tag, error) {
 		})
 	}
 
-	r.sortTags(tags)
-	r.assignPreviousAndNextTag(tags)
+	switch r.sortBy {
+	case "date":
+		r.sortTags(tags)
+		r.assignPreviousAndNextTag(tags)
+	case "semver":
+		filterSemVerTags(&tags)
+		r.sortTagsBySemver(tags)
+		r.assignPreviousAndNextTag(tags)
+	}
 
 	return tags, nil
 }
 
+func filterSemVerTags(tags *[]*Tag) {
+	// filter out any non-semver tags
+	for i, t := range *tags {
+		// remove leading v, since its so
+		// common.
+		name := t.Name
+		if strings.HasPrefix(name, "v") {
+			name = strings.TrimPrefix(name, "v")
+		}
+
+		// attempt semver parse, if not successful
+		// remove it from tags slice.
+		if _, err := semver.NewVersion(name); err != nil {
+			*tags = append((*tags)[:i], (*tags)[i+1:]...)
+		}
+	}
+}
```

Comment:
For consistnecy, lets use the receiver type of `(*tagReader)` to make this a struct method. There are other methods that modify the `[]*Tag` slice in place for this struct.

```suggestion
func (*tagReader) filterSemVerTags(tags []*Tag) {
	// filter out any non-semver tags
	for i, t := range tags {
		// remove leading v, since its so
		// common.
		name := t.Name
		if strings.HasPrefix(name, "v") {
			name = strings.TrimPrefix(name, "v")
		}

		// attempt semver parse, if not successful
		// remove it from tags slice.
		if _, err := semver.NewVersion(name); err != nil {
			tags = append((tags)[:i], (tags)[i+1:]...)
		}
	}
}
```

Reply:
Won't this not work? In this example, we are modifying a copy of the tags array, we aren't modifying the original one.

if I run tests with this method, we never replace the original slice header, and the non-semver is included, passed to the semver parse method, and we get a panic. 
```
❯ go test ./...
--- FAIL: TestTagReader (0.00s)
panic: hoge_fuga is not in dotted-tri format [recovered]
	panic: hoge_fuga is not in dotted-tri format

goroutine 160 [running]:
testing.tRunner.func1.2(0x7cf000, 0xc0000915a0)
	/usr/local/go/src/testing/testing.go:1144 +0x332
testing.tRunner.func1(0xc000582300)
	/usr/local/go/src/testing/testing.go:1147 +0x4b6
panic(0x7cf000, 0xc0000915a0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/coreos/go-semver/semver.Must(...)
	/home/louis/git/gopath/pkg/mod/github.com/coreos/go-semver@v0.3.0/semver/semver.go:65
github.com/coreos/go-semver/semver.New(0xc00048abc0, 0x9, 0x4)
	/home/louis/git/gopath/pkg/mod/github.com/coreos/go-semver@v0.3.0/semver/semver.go:49 +0x77
github.com/git-chglog/git-chglog.(*tagReader).sortTagsBySemver.func1(0x5, 0x4, 0x1)
	/home/louis/git/go/git-chglog/tag_reader.go:159 +0x119
sort.insertionSort_func(0xc000125aa8, 0xc0004967a0, 0x0, 0x6)
	/usr/local/go/src/sort/zfuncversion.go:12 +0xab
sort.quickSort_func(0xc000125aa8, 0xc0004967a0, 0x0, 0x6, 0x6)
	/usr/local/go/src/sort/zfuncversion.go:158 +0x1f6
sort.Slice(0x7b5220, 0xc00033e3f0, 0xc000125aa8)
	/usr/local/go/src/sort/slice.go:20 +0xe8
github.com/git-chglog/git-chglog.(*tagReader).sortTagsBySemver(0xc000125f00, 0xc0002e2b00, 0x6, 0x8)
	/home/louis/git/go/git-chglog/tag_reader.go:156 +0xa5
github.com/git-chglog/git-chglog.(*tagReader).ReadAll(0xc000125f00, 0x0, 0xc000329220, 0x7b5220, 0xc00033e3d8, 0x0)
	/home/louis/git/go/git-chglog/tag_reader.go:83 +0x4a5
github.com/git-chglog/git-chglog.TestTagReader(0xc000582300)
	/home/louis/git/go/git-chglog/tag_reader_test.go:123 +0x117f
testing.tRunner(0xc000582300, 0x874da8)
	/usr/local/go/src/testing/testing.go:1194 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1239 +0x2b3
FAIL	github.com/git-chglog/git-chglog	1.104s
ok  	github.com/git-chglog/git-chglog/cmd/git-chglog	(cached)
FAIL
```

Reply:
Was the call to `r.filterSemVerTags` changed from `filterSemVerTags(&tags)` to `r.filterSemVerTags(tags)` when this test was run?

I think that may be the issue.

Reply:
@ldelossa my bad. I was confusing the way that the other methods employ the `sort.Slice` method. You are correct that you would need to pass a pointer in order to filter the slice in place.

Reply:
Np, honestly im a little confused why thats not necessary for sort.Slice? Maybe because it *only* changes the underlying array and not the slice header? 

But regardless, ill def update to taking a receiver.

Reply:
Yeah, it's something I want to read up on. It feels odd. 

Thanks for following up on this!